doc/man3/flux_future_promise.adoc

```diff
@@ -0,0 +1,163 @@
+flux_future_promise(3)
+======================
+:doctype: manpage
+
+
+NAME
+----
+flux_future_promise, flux_future_set_result, flux_future_set_error, flux_future_aux_get, flux_future_aux_set, flux_future_set_flux, flux_future_get_flux - support methods for future wrappers
+
+
+SYNOPSIS
+--------
+ #include <flux/core.h>
+
+ typedef void (*flux_promise_f)(flux_future_t *f,
+                                flux_reactor_t *r, void *arg);
+
+ void flux_future_promise (flux_future_t *f,
+                           flux_promise_f cb, void *arg);
+
+ void flux_future_set_result (flux_future_t *f,
+                              void *result, flux_free_f free_fn);
+
+ void flux_future_set_error (flux_future_t *f, int errnum);
+
+ void *flux_future_aux_get (flux_future_t *f, const char *name);
+
+ int flux_future_aux_set (flux_future_t *f, const char *name,
+                          void *aux, flux_free_f destroy);
+
+ void flux_future_set_flux (flux_future_t *f, flux_t *h);
+
+ flux_t *flux_future_get_flux (flux_future_t *f);
+
+
+DESCRIPTION
+-----------
+
+A Flux future represents some activity that may be completed with reactor
+watchers and/or message handlers.  It is intended to be wrapped
+by other classes that perform specific activities, providing a common idiom
+for synchronizing on the result.  Typically such a class provides
+class-specific wrappers for `flux_future_create()` and `flux_future_get()`,
+and implements a class-specific promise callback.
+
+`flux_future_promise()` registers a promise callback.  The promise sets
+up class-specific watchers on the reactor to fulfill the promise as
+asynchronous events occur.  The watchers must eventually call
+`flux_future_set_result()` or `flux_future_set_error()` to fulfill
+the promise and complete the future.
+
+`flux_future_set_result()` assigns an opaque _result_ value to the
+future, with optional destructor.  NULL is a valid result.
+The assignment completes the future and triggers synchronization,
+after which `flux_future_get()` can be used to obtain the _result_
+value without blocking.  The destructor is called when the future is
+destroyed.
+
+`flux_future_set_error()` similarly completes the future and triggers
+synchronization, but instead of assigning a result, it assigns an
+errno value, after which `flux_future_get()` returns -1 with errno
+set to this value.
+
+`flux_future_aux_set()` attaches an object _aux_, with optional
+destructor _destroy_, to the future under an optional _name_.
+If the _name_ is NULL, the object is stored anonymously.
+`flux_future_aux_get()` retrieves an object by _name_.
+Anonymous objects cannot be retrieved.  All aux objects are destroyed
+when the future is destroyed.
+
```

Comment:
There was this, but I'll see what I can do to make that more clear.  The idea was not to make the objects retrievable, just to get a destructor to run during future destruction.

> Since the promise may be called in either reactor context (at most once each),
and is unaware of which context that is, it should take care when managing
any context-specific state not to overwrite the state from a prior call.
The ability to attach objects with destructors anonymously to the future
with `flux_future_aux_set()` may be useful for managing the life cycle
of reactor watchers and message handlers created by promises.